gdal/alg/marching_squares/polygon_ring_appender.h

```diff
@@ -129,49 +131,69 @@ class PolygonRingAppender
 
     void addLine( double level, LineString& ls, bool )
     {
-        Ring r;
-        r.points.swap( ls );
-        rings_[level].push_back( std::move(r) );
+        // Create a new ring from the LineString
+        Ring newRing;
+        newRing.points.swap( ls );
+        auto &levelRings = rings_[level];
+        // This queue holds the rings to be checked
+        std::deque<Ring*> queue;
+        std::transform(levelRings.begin(),
+                       levelRings.end(),
+                       std::back_inserter(queue),
+                       [](Ring &r) {
+                           return &r;
+                       });
+        Ring *parentRing = nullptr;
+        while( !queue.empty() ) {
+            Ring *curRing = queue.front();
+            queue.pop_front();
+            if ( newRing.isIn(*curRing) ) {
+                // We know that there should only be one ring per level that we should fit in,
+                // so we can discard the rest of the queue and try again with the children of this ring
+                parentRing = curRing;
+                queue.clear();
+                std::transform(curRing->interiorRings.begin(),
+                            curRing->interiorRings.end(),
+                            std::back_inserter(queue),
+                            [](Ring &r) {
+                                return &r;
+                            });
+            }
+        }
+        if ( parentRing == nullptr ) {
+            // Then we never found anything and this must be a new top level ring for this level
+            levelRings.push_back(newRing);
+        } else {
+            // We found a valid parent, so we need to:
+            // 1. Find all the inner rings of the parent that are inside the new ring
+            auto trueGroupIt = std::partition(
+                parentRing->interiorRings.begin(),
+                parentRing->interiorRings.begin(),
```

Comment:
Oops, this is a typo, should be `parentRing->interiorRings.end()`, which (I think?) turns this whole section into pretty much a no-op, as the partitioned result will always be of size 0. This might be causing my test failures, investigating.